### PR TITLE
Repo tidy up

### PR DESCRIPTION
**Pull request by danpat**
This is a repo tidy-up PR.  It'll cover:

  - standardizing code formatting/style and enforcing that
  - measuring test coverage and improving it
  - running some sanitizers over the code to catch unknown bugs
  - updating docs

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ---
+# Mapbox.Variant C/C+ style
```
Comment (daniel-j-h) [inline .clang-format]: Why the change?

File: .clang-tidy [hunk 2]
```diff
@@ -0,0 +1,28 @@
+---
```
Comment (daniel-j-h) [inline .clang-tidy]: For which clang-tidy version is this for?

File: .travis.yml [hunk 3]
```diff
@@ -5,38 +5,159 @@ env:
...
     - secure: "JV6xOo1YVqUUlsa/za7BGuYZj//y5vg7Dgcn6v0MJzF+ZxIM5maGh6iA3mhiCIUetHVwwlqjJg8ZaZ8vFcWRcx27+591xgTSJWrPhgnbqcNupDCGoexiqUUzfgOtkuDowmRp4wz2e5uUpox4Gm6gmJkEt1nUs+/gsk7fjDoChSTGgJ2eASiLgzY2Hg38iT5IiZMx//FWw0uQnYCtPR0vtGYdgT8aSMXnVSsFqvpUJBUnQj4R93eJbQiXUN6SCRpjvd536JOgd+ud+zm7QTvzmRKSgag31XDKNjD/1SK3zzhkfNL/1T7Kr3rCdluxa0RO2c3eLiN6rkjelVtFcRvj7zwJvZ2BJ5WLWILtvbvvCmNzI43v+8RandYiWm8CWxVwUt3Cw+xzG0DblO+TCT8P9DipljI5zOeEBh6Tc920zb0P45E5jnxu6oxS2sOV/K2oyFu5152JXI4u24xVxMeVd9cxOzd4Jpxn5pe/m7RPeEDTpHA28RBTjA98kysR3aUBsfF5I/IrxkT8e9XnMuAOGaVgaAzGRAjL481H8CMIkQT14hO3+jqPKcn1scgDOGxw+Ijujdix9BtVt5/pEokfXYd2KXBmgDAPTBqfgkabPDXoNv0iXDdkQVl0924s0Jq/Io1DbjIIS+zezrPOwryVBsPjRYfWsjwl9Nv1KjK+yeQ="
 
+
+# enable c++11/14 builds
 addons:
   apt:
-    sources: [ 'ubuntu-toolchain-r-test', 'llvm-toolchain-trusty-4.0' ]
-    packages: [ 'libstdc++-4.9-dev', 'clang-4.0' ]
+    sources: [ 'ubuntu-toolchain-r-test' ]
```
Comment (daniel-j-h) [inline .travis.yml]: Why removing the llvm toolchain source and clang?

File: .travis.yml [hunk 4]
```diff
@@ -5,38 +5,159 @@ env:
...
+      env: BUILDTYPE=release
       node_js: 6
-      sudo: false
-      env: CC='clang-4.0' CXX='clang++-4.0'
+    # linux publishable node v6/debug
     - os: linux
-      node_js: 8
-      sudo: false
-      env: CC='clang-4.0' CXX='clang++-4.0'
+      env: BUILDTYPE=debug
```
Comment (daniel-j-h) [inline .travis.yml]: Debug builds need debug libortools otherwise they won't help at all

File: binding.gyp [hunk 5]
```diff
@@ -1,68 +1,123 @@
...
-    'includes': [ 'common.gypi' ],
-    'variables': {
-      # Includes we don't want warnings for.
+  # https://github.com/springmeyer/gyp/blob/master/test/make_global_settings/wrapper/wrapper.gyp
+  'make_global_settings': [
+    ['CXX', '<(module_root_dir)/mason_packages/.link/bin/clang++'],
+    ['CC', '<(module_root_dir)/mason_packages/.link/bin/clang'],
+    ['LINK', '<(module_root_dir)/mason_packages/.link/bin/clang++'],
+    ['AR', '<(module_root_dir)/mason_packages/.link/bin/llvm-ar'],
+    ['NM', '<(module_root_dir)/mason_packages/.link/bin/llvm-nm']
```
Comment (daniel-j-h) [inline binding.gyp]: Why do these compilation settings depend on mason? Folks should be able to compile with their system compiler, no? Can we set paths externally and keep the build system agnostic to mason?

File: binding.gyp [hunk 6]
```diff
@@ -1,68 +1,123 @@
...
+      # Use this variable to silence warnings from mason dependencies and from NAN
+      # It's a variable to make easy to pass to
+      # cflags (linux) and xcode (mac)
       'system_includes': [
-        '-isystem <(module_root_dir)/<!(node -e \'require("nan")\')',
-        '-isystem <(module_root_dir)/mason_packages/.link/include',
-        '-isystem <(module_root_dir)/mason_packages/.link/include/or-tools'
+        "-isystem <(module_root_dir)/<!(node -e \"require('nan')\")",
+        "-isystem <(module_root_dir)/mason_packages/.link/include/",
+        "-isystem <(module_root_dir)/mason_packages/.link/include/or-tools"
```
Comment (daniel-j-h) [inline binding.gyp]: Same here

File: binding.gyp [hunk 7]
```diff
@@ -1,68 +1,123 @@
...
+            '-lz',
+          ],
+      },
+      'sources': [
+          './src/main.cc',
+          './src/tsp.cc',
+          './src/vrp.cc',
+      ],
+      'ldflags': [
+        '-Wl,-z,now',
```
Comment (daniel-j-h) [inline binding.gyp]: Why?

File: binding.gyp [hunk 8]
```diff
@@ -1,68 +1,123 @@
...
+        }]
+      ],
+      'cflags_cc': [
+          '-std=c++14',
+          '<@(system_includes)',
+          '<@(compiler_checks)'
+      ],
+      'xcode_settings': {
+        'OTHER_LDFLAGS':[
+          '-Wl,-bind_at_load'
```
Comment (daniel-j-h) [inline binding.gyp]: Why?

File: common.gypi [hunk 9]
```diff
@@ -2,14 +2,15 @@
...
     'cflags_cc' : [
-      '-std=c++14',
-      '-Wall',
-      '-Wextra',
-      '-ffunction-sections -fdata-sections',
-      '-D_GLIBCXX_USE_CXX11_ABI=0',
-    ],
-    'ldflags': [
-      '-Wl,--gc-sections'
+      '-std=c++11',
```
Comment (daniel-j-h) [inline common.gypi]: Why C++11 here but above we use C++14?

File: common.gypi [hunk 10]
```diff
@@ -36,12 +37,20 @@
         'defines': [
           'NDEBUG'
         ],
+        'cflags': [
+         '-flto'
```
Comment (daniel-j-h) [inline common.gypi]: Why? How will this help when libortools is not compiled in lto format?

File: package.json [hunk 11]
```diff
@@ -18,15 +18,19 @@
   "scripts": {
     "install": "node-pre-gyp install --fallback-to-build",
     "clean": "node-pre-gyp clean",
-    "test": "tap -Rspec test/*.js"
+    "test": "tape test/*.test.js"
```
Comment (daniel-j-h) [inline package.json]: Why the change?

File: package.json [hunk 12]
```diff
@@ -18,15 +18,19 @@
...
+    "nan": "^2.10",
     "node-pre-gyp": "^0.6.34"
   },
   "devDependencies": {
     "aws-sdk": "^2.42.0",
-    "tap": "^10.3"
+    "tape": "^4.5.1",
+    "d3-queue": "^3.0.1",
+    "minimist": "~1.2.0",
+    "bytes": "^2.4.0"
```
Comment (daniel-j-h) [inline package.json]: Why do we need these now?

File: package.json [hunk 13]
```diff
@@ -18,15 +18,19 @@
   "scripts": {
     "install": "node-pre-gyp install --fallback-to-build",
     "clean": "node-pre-gyp clean",
-    "test": "tap -Rspec test/*.js"
+    "test": "tape test/*.test.js"
   },
   "dependencies": {
-    "nan": "^2.6",
+    "mason-js-sdk": "~0.1.4",
```
Comment (daniel-j-h) [inline package.json]: Should this be a dev dependency?

File: scripts/clang-format.sh [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+Return `1` if there are files to be formatted, and automatically formats them.
+
+Returns `0` if everything looks properly formatted.
+
+'
+
+PATH_TO_FORMAT_SCRIPT="$(pwd)/mason_packages/.link/bin/clang-format"
+
+# Run clang-format on all cpp and hpp files in the /src directory
+find src/ -type f -name '*.hpp' -o -name '*.cpp' -o -name '*.h' -o -name '*.cc' \
```
Comment (daniel-j-h) [inline scripts/clang-format.sh]: We only have cc and h files

File: scripts/clang-format.sh [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+Runs clang-format on the code in src/
+
+Return `1` if there are files to be formatted, and automatically formats them.
+
+Returns `0` if everything looks properly formatted.
+
+'
+
+PATH_TO_FORMAT_SCRIPT="$(pwd)/mason_packages/.link/bin/clang-format"
```
Comment (daniel-j-h) [inline scripts/clang-format.sh]: Why do these scripts depend on mason? mason should put clang-format into PATH and these scripts should not all have to know about mason.

File: scripts/clang-tidy.sh [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+Runs clang-tidy on the code in src/
+Return `1` if there are files automatically fixed by clang-tidy.
+Returns `0` if no fixes by clang-tidy.
+TODO: should also return non-zero if clang-tidy emits warnings
+or errors about things it cannot automatically fix. However I cannot
+figure out how to get this working yet as it seems that clang-tidy
+always returns 0 even on errors.
+'
+
+PATH_TO_CLANG_TIDY_SCRIPT="$(pwd)/mason_packages/.link/share/run-clang-tidy.py"
```
Comment (daniel-j-h) [inline scripts/clang-tidy.sh]: Same; also what is `run-clang-tidy.py`?

File: scripts/clang-tidy.sh [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+# make sure that run-clang-tidy.py can find the right clang-tidy
+export PATH=$(pwd)/mason_packages/.link/bin:${PATH}
+
+# build the compile_commands.json file if it does not exist
```
Comment (daniel-j-h) [inline scripts/clang-tidy.sh]: Why? Why do we need the compile commands database?

File: scripts/generate_compile_commands.py [hunk 18]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env python
```
Comment (daniel-j-h) [inline scripts/generate_compile_commands.py]: Should be `python2` or `python3`

File: scripts/generate_compile_commands.py [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+    compile_commands = []
+    for line in sys.stdin.readlines():
+        if TOKEN_DENOTING_COMPILED_FILE in line:
+            match = matcher.match(line)
+            compile_commands.append({
+                "directory": build_dir,
+                "command": line.strip(),
+                "file": os.path.normpath(os.path.join(build_dir,match.group(2)))
+            })
+    print json.dumps(compile_commands,indent=4)
```
Comment (daniel-j-h) [inline scripts/generate_compile_commands.py]: This only works with Python2, in Python3 print is a function.

File: scripts/sanitize.sh [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+    ./node_modules/.bin/mason-js install binutils=2.30 --type=compiled
+    ./node_modules/.bin/mason-js link binutils=2.30 --type=compiled
+    export PATH=$(pwd)/mason_packages/.link/bin:${PATH}
+fi
+export MASON_LLVM_RT_PRELOAD=$(pwd)/$(ls mason_packages/.link/lib/clang/*/lib/*/libclang_rt.asan*${SHARED_LIB_EXT})
+SUPPRESSION_FILE="/tmp/leak_suppressions.txt"
+echo "leak:__strdup" > ${SUPPRESSION_FILE}
+echo "leak:v8::internal" >> ${SUPPRESSION_FILE}
+echo "leak:node::CreateEnvironment" >> ${SUPPRESSION_FILE}
+echo "leak:node::Init" >> ${SUPPRESSION_FILE}
```
Comment (daniel-j-h) [inline scripts/sanitize.sh]: Why these suppressions? And should this be a suppression file for this repo and not implicitly generated behind the user's back?

File: scripts/sanitize.sh [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+make debug
+export ASAN_OPTIONS=fast_unwind_on_malloc=0:${ASAN_OPTIONS}
+if [[ $(uname -s) == 'Darwin' ]]; then
+    # NOTE: we must call node directly here rather than `npm test`
+    # because OS X blocks `DYLD_INSERT_LIBRARIES` being inherited by sub shells
+    # If this is not done right we'll see
+    #   ==18464==ERROR: Interceptors are not working. This may be because AddressSanitizer is loaded too late (e.g. via dlopen).
+    #
+    DYLD_INSERT_LIBRARIES=${MASON_LLVM_RT_PRELOAD} \
+      node node_modules/.bin/tape test/*test.js
```
Comment (daniel-j-h) [inline scripts/sanitize.sh]: This needs to be in sync with the package json's test command, no?

File: test/tsp.test.js [hunk 22]
```diff
@@ -43,7 +43,7 @@ tap.test('Test TSP', function(assert) {
   TSP.Solve(searchOpts, function (err, solution) {
     assert.ifError(err, 'Solution can be found');
 
-    assert.type(solution, Array, 'Solution is Array of locations');
+    assert.ok(solution instanceof Array, 'Solution is Array of locations');
```
Comment (daniel-j-h) [inline test/tsp.test.js]: Why?

File: .clang-tidy (hunk 2 above)
Comment (daniel-j-h) [reply .clang-tidy]: It's not solved, that's the problem :( same with clang-format where versions are _slightly_ incompatible (that's why we fix a specific version in osrm for all devs).

Comment (springmeyer) [reply .clang-tidy]: Overall @danpat is using `node-cpp-skel` latest here, so I'll share context below on the decisions in node-cpp-skel. If you want to change/improve/debate `node-cpp-skel` design decisions feel free to open tickets over at https://github.com/mapbox/node-cpp-skel/issues.

Specifically node-cpp-skel framework installs its own version of clang-tidy/clang-format as a way to provide a consistent version for devs (when you type `make`). If you want to change the clang-tidy version (and are using the framework from node-cpp-skel like @danpat is doing here) you now control the version via the `mason-versions.ini` file.

File: .travis.yml (hunk 3 above)
Comment (springmeyer) [reply .travis.yml]: Because @danpat copied over the `.travis.yml` from node-cpp-skel which moves control of the compiler to `mason-js`, which installs a consistent version of clang++ via mason.

If anyone wants to add back jobs that use different compilers, feel free to do that. However node-cpp-skel only offers a template that uses clang++ via mason and g++ via 'ubuntu-toolchain-r-test' currently.

File: .travis.yml (hunk 4 above)
Comment (springmeyer) [reply .travis.yml]: In my experience there is still value to `Debug` builds (even if an external library is not also built with symbols) since `Debug` builds might trigger more warnings or uncover bugs in the node specific code that is compiled.

But yes, if we want a debug version of `libortools`, we could package that in mason (or this debug job could compile it from source on the fly?).

File: binding.gyp (hunk 5 above)
Comment (springmeyer) [reply binding.gyp]: Yes, you can use an alternative compiler 👍 These are just defaults. See https://github.com/mapbox/node-cpp-skel/blob/master/README.md#customizing-the-compiler-toolchain for details.

File: scripts/generate_compile_commands.py (hunk 18 above)
Comment (springmeyer) [reply scripts/generate_compile_commands.py]: Please provide a PR to https://github.com/mapbox/node-cpp-skel if you think this needs fixed. We use `#!/usr/bin/env python` currently since this works on: a) latest OS X and b) ubuntu precise/trusty.

Does this break with latest ubuntu? How can we support all systems here?

File: scripts/generate_compile_commands.py (hunk 19 above)
Comment (springmeyer) [reply scripts/generate_compile_commands.py]: I've not tried to get python3 support working in node-cpp-skel because `gyp` itself only supports python2. So, fixing this is probably not worth it. The larger issue is probably that we should move node-cpp-skel to use `cmake` and `node-cmake`, that way we'd be able to drop the `python` dependency completely. And this file would be able to go away since we could use `cmake`s built-in compile_commands support.

File: .clang-tidy (hunk 2 above)
Comment (danpat) [reply .clang-tidy]: 👍 You can either use `make tidy` or `./scripts/clang-tidy.sh` to run the project-preferred version of `clang-tidy`

If someone wants to use a system version of `clang-tidy`, well, there's not much we can do to stop them.  Our CI builds will only use the invocations above though, so ultimately, CI is our enforcer on this.

File: .travis.yml (hunk 3 above)
Comment (danpat) [reply .travis.yml]: There is a `# g++ build (default builds all use clang++)` build which adds `g++6` support, and shows how to override the Makefile default compiler paths.  Using system installed compilers is still fairly easy, just no longer the default.

File: scripts/sanitize.sh (hunk 21 above)
Comment (springmeyer) [reply scripts/sanitize.sh]: It does. Ticketed thinking about avoiding this at https://github.com/mapbox/node-cpp-skel/issues/122

File: test/tsp.test.js (hunk 22 above)
Comment (springmeyer) [reply test/tsp.test.js]: This one is not related to upgrading to use latest node-cpp-skel, so I'll let @danpat answer.

Comment (danpat) [reply test/tsp.test.js]: It's only semi-related.  `node-cpp-skel` suggests using `tape`, but this project was using `tap` for testing - the change above is because `tape` doesn't support the `.type()` test function.

File: binding.gyp (hunk 8 above)
Comment (danpat) [reply binding.gyp]: This is necessary for dynamic libraries to load inside the nodejs environment.  The previous `binding.gyp` had the same setting (see up a few lines), it's just been moved to a new location.

File: common.gypi (hunk 9 above)
Comment (danpat) [reply common.gypi]: The `common.gypi` is a reusable set of defaults for `node-gyp` projects from the `node-cpp-skel` project.  Values here are intended to be overridden by more specific values in `binding.gyp`, which is what this code now does.

File: scripts/clang-tidy.sh (hunk 16 above)
Comment (danpat) [reply scripts/clang-tidy.sh]: https://github.com/llvm-mirror/clang-tools-extra/blob/master/clang-tidy/tool/run-clang-tidy.py

It's bundled with `clang-tidy`, and does 1) parallel scanning of files, and 2) reads files to check from `compile_commands.json`

File: scripts/clang-tidy.sh (hunk 17 above)
Comment (danpat) [reply scripts/clang-tidy.sh]: See above - needed by `run-clang-tidy.py`

File: scripts/generate_compile_commands.py (hunk 18 above)
Comment (danpat) [reply scripts/generate_compile_commands.py]: PR is done upstream, and I've corrected this script to work with both Python2 and 3.

More details on current state of what `python` means is detailed in that PR: https://github.com/mapbox/node-cpp-skel/pull/120#issuecomment-384404925

File: scripts/generate_compile_commands.py (hunk 19 above)
Comment (danpat) [reply scripts/generate_compile_commands.py]: Fixed.

File: src/vector.h [hunk 23]
```diff
@@ -8,13 +8,13 @@ template <typename T> class Vector {
 public:
   using Value = T;
 
-  Vector() = default;
-  Vector(std::int32_t n) { data.resize(n); }
+  Vector() : data{} {}
+  Vector(std::int32_t n) : data(static_cast<std::size_t>(n)) {}
```
Comment (artemp) [inline src/vector.h]:
```c++
Vector<int> v(-1);
```

```bash
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Aborted (core dumped)
```
Just caught my eye ^. Does signed int really needed? 
Why do we need this `std::vector` wrapper at all ?

Comment (daniel-j-h) [reply src/vector.h]: I think we started out with a matrix wrapper https://github.com/mapbox/node-or-tools/blob/4364bd69e35c5e25504f0d6dd2c9d1ac829f5a57/src/matrix.h and then just went with it.

Re. your question for signed ints: this project uses signed ints everywhere and we consider unsigned ints in APIs like vector or for anything else then bitmasks / bytes a design mistake.

Your `vector v(-1)` is a good example for this: the signed `-1` integer literal gets converted to `numeric_limits<size_t>::max()` and then the vector constructor has no way of asserting `n >= 0` for the `vector(n)` constructor. Similarly in other situations with unsigned ints the sanitizers can't catch overflows or other issues easily.

It's unfortunate that the stdlib is unsigned in general but it's a mistake we don't have to repeat.

Comment (artemp) [reply src/vector.h]: Ok, thanks for the background.
> It's unfortunate that the stdlib is unsigned in general but it's a mistake we don't have to repeat.
Do you have some links etc ? I'd love to catch up on this as it comes up quite a lot ^

Comment (daniel-j-h) [reply src/vector.h]: I _think_ there were some talks and discussions at C++ conferences about this; some from the Google folks in the context of their sanitizers and detecting these issues. And this discussion is going on in the larger C++ community (irc, mailinglists, twitter, reddit, ...) but I don't have good resources to point you at of the top of my head, sorry.

Comment (danpat) [reply src/vector.h]: @daniel-j-h So, part of the cleanup require here was fixing a bunch of signed/unsigned cast warnings because we've added `-Wconversion` to the build flags.  There were a *tonne* of places where signed/unsigned values were being converted, a couple very dubiously.

I've fixed all the warnings, but to me, the API feels a bit off - there are very few places where we actually accept negative numbers - why would we make these parameters signed at all?

File: common.gypi (hunk 10 above)
Comment (danpat) [reply common.gypi]: I guess it won't?  Like other changes, this is a default pulled in via `common.gypi`.  Assuming `-flto` does no harm if the linked libs aren't lto capable, then I'm inclined to just leave this here.

File: package.json (hunk 11 above)
Comment (danpat) [reply package.json]: The node-cpp-skel (which this project was first based on) framework assumes `tape` instead of `tap`.  They're much the same, but using `tape` will make life simpler in the future if we pull in updates from `node-cpp-skel`.

File: src/vector.h (hunk 23 above)
Comment (daniel-j-h) [reply src/vector.h]: See conversation and arguments above. Unsigneds silently wrap around and you can't detect errors with them. Unsigneds should only be used for bitmasks / bytes.

Comment (daniel-j-h) [reply src/vector.h]: @danpat unsigned wraps around - that's exactly the problem, you can't detect an issue when it overflows. Signed overflow is undefined which allows the sanitizers (or other checkers) to catch it. You usually never want the overflow semantics; a `vector(-1)` should error out and not try to allocate `numeric_limits<size_t>::max()` memory.

**Comment by danpat** [comment]
Stale
